File: 144-binary-tree-preorder-traversal/binary-tree-preorder-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
// Preorder Traversal ROOT, LEFT, RIGHT
// Recursive
var preorderTraversal = function(root) {
    if (!root) return []
    return [root.val, ...preorderTraversal(root.left), ...preorderTraversal(root.right)]
};

File: 144-binary-tree-preorder-traversal/binary-tree-preorder-traversal.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func preorderTraversal(_ root: TreeNode?) -> [Int] {
        var curr: TreeNode? = root, stack: [TreeNode] = [] 
        var preorder: [Int] = []
        while curr != nil || !stack.isEmpty {
            if curr == nil { curr = stack.removeLast()}
            preorder.append(curr!.val)
            if var right = curr?.right { stack.append(right)}
            curr = curr?.left
        }   
        return preorder
    }
}